src/librustc_mir/transform/const_prop.rs

```diff
@@ -658,10 +657,73 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
         if rvalue.needs_subst() {
             return None;
         }
+        // self.use_ecx(|this| {
```

Comment:
You can do the mir-opt-level test right here (by reading `self.tcx.sess.opts.debugging_opts.mir_opt_level`), doing `eval_rvalue_with_identities` if `>= 3` and `eval_rvalue_into_place` otherwise.

You'll need to add `// compile-flags: -Zmir-opt-level=3` to your tests to make them actually do something after that.